src/fromager/dependency_graph.py

```diff
@@ -352,6 +352,19 @@ def get_nodes_by_name(self, req_name: str | None) -> list[DependencyNode]:
     def get_root_node(self) -> DependencyNode:
         return self.nodes[ROOT]
 
+    def get_top_level_requirement(self, node: DependencyNode) -> Requirement | None:
+        """Get the top-level requirement specification for a node.
+
+        For packages that were specified as top-level requirements (e.g., with git URLs),
+        this returns the original requirement specification. Returns None if the node
+        is not a direct child of ROOT.
+        """
+        root = self.get_root_node()
+        for edge in root.children:
+            if edge.destination_node.key == node.key:
```

Comment:
I think you can say `edge.destination_node is node`?

You could also look through `node.parents` for an edge with a destination node that is the root node. I don't know which is likely to be more efficient in terms of iterative over fewer items on average.